### ISSUE TITLE
Add command to allow kernel-restart without confirmation dialog

### ISSUE THREAD

**Issue by javabrett** [jupyterlab/jupyterlab#6420]
When working with notebooks in a development-context, perhaps coding underlying APIs and libraries, it is often desirable to completely restart the running kernel, so that all changed code is loaded and the context is cleared.

Currently when doing that, a confirmation dialog has to be accepted each time.  In development mode it would be useful if such a dialog could be bypassed, i.e. kernel-restart with one click or keystroke.

**Comment by vidartf** [comment]
A possible solution could depend on https://github.com/jupyterlab/jupyterlab/issues/5617.

**Comment by saulshanabrook** [comment]
I would support this. I often restart and run all when I am dealing with stateful computations and the dialogue box gets in the way.

**Comment by BoPeng** [comment]
To me restarting kernel is a major operation as rerunning my notebook might take hours, so I would oppose a shortcut that restarts my kernel without confirmation. 

Also there is already a widely used pattern for such issues:

1. The first time the dialog appears (triggered by the shortcut), it has an option says 
   `[x] Show this dialog next time`
2. Unselect the box if you do not want a confirmation box next time.
3. Somewhere in the preferences there is an option to bring this dialog back.

**Comment by vidartf** [comment]
@BoPeng See #5617 :)

**Comment by CmdQ** [comment]
In my work I have to restart the kernel about 30 times a day and it's plain annoying to always have to do that click. But otherwise it won't find the new code that found it's way into the repo.

It wouldn't have to be #5617, which brings more work for all dialogs. It could also be a setting in the options like “developer mode” or similar.

**Comment by jasongrout** [comment]
FYI, you can just press "enter" to confirm the restart and make the dialog go away. You might find that faster than clicking, at least until there is a better way to optionally get rid of the dialog entirely.

**Comment by CmdQ** [comment]
Yes that's true of course. Anyway, I'd rather have the I know what I'm doing mode. 😎

**Comment by jasongrout** [comment]
> Yes that's true of course. Anyway, I'd rather have the I know what I'm doing mode. 😎

Same here. Just trying to ease the pain in the short term.

As with many issues, people are welcome to work on this and move it forward!

**Comment by calebmarchent** [comment]
Just like to add my "vote" to wanting this feature.  When hiving off the logic into python libraries, so that I can keep the notebook itself succinct end up needing to reload each time I change the imported .py module, so this option would be really helpful.

**Comment by mvoelk** [comment]
Yes please!

**Comment by fahadijaz** [comment]
This works perfectly fine to resolve this issue.

from IPython.core.display import HTML
HTML("<script>Jupyter.notebook.kernel.restart()</script>")

as stated over here 
https://stackoverflow.com/questions/37472274/jupyter-disable-restart-kernel-warning/
by kingbase

**Comment by krassowski** [comment]
@fahadijaz thanks for the link. I am afraid this does not do anything in Jupyter**Lab** though.

As for this issue, could we just have an optional argument `ask: boolean = false` added to the `kernelmenu:restart` command, which then could be bound to a shortcut, or invoked from ipylab (see https://discourse.jupyter.org/t/is-there-magic-to-restart-kernel-change-kernel/10763/2).

**Comment by fahadijaz** [comment]
That would be a useful and practical solution.

**Comment by banshee** [comment]
> FYI, you can just press "enter" to confirm the restart

Not for me - Chrome Version 100.0.4896.60 (Official Build) (64-bit) Ubuntu.  Pressing enter and pressing escape both just dismiss the dialog.  Hitting tab to switch back and forth between the two buttons does work, but even when Restart is clearly selected, the enter button just dismisses the dialog.  Haven't found a combo that actually presses the button; shift-enter, ctrl-enter, etc all just dismiss.

**Comment by krassowski** [comment]
> Not for me

@banshee in which JupyterLab version do you see this issue?

**Comment by banshee** [comment]
> @banshee in which JupyterLab version do you see this issue?

Jupyter Server 1.13.5, Lab about box says Version 3.2.9.  Lots of extensions loaded, and I did disable AdBlock for the site.  No interesting messages on the Chrome console, although I do see this message for both hitting return and clicking the button:

```
Rxjs.ipynb:1 Uncaught (in promise) Error: The message port closed before a response was received.
react_devtools_backend.js:3973 Connection lost, reconnecting in 0 seconds.
overrideMethod @ react_devtools_backend.js:3973
_reconnect @ default.jss:1245
reconnect @ default.jss:526
restart @ default.jss:496
await in restart (async)
restartKernel @ sessioncontext.jss:319
restart @ sessioncontext.jss:788
await in restart (async)
restartKernel @ index.jss:1966
(anonymous) @ index.jss:560
e.execute @ index.es6.jss:372
e._executeKeyBinding @ index.es6.jss:532
e.processKeydownEvent @ index.es6.jss:471
e.evtKeydown @ index.es6.jss:355
e.handleEvent @ index.es6.jss:311
react_devtools_backend.js:3973 Connection lost, reconnecting in 0 seconds.
overrideMethod @ react_devtools_backend.js:3973
_reconnect @ default.jss:1245
reconnect @ default.jss:526
restart @ default.jss:496
await in restart (async)
restartKernel @ sessioncontext.jss:319
restart @ sessioncontext.jss:788
await in restart (async)
restartKernel @ index.jss:1966
(anonymous) @ index.jss:560
e.execute @ index.es6.jss:372
e._executeKeyBinding @ index.es6.jss:532
e.processKeydownEvent @ index.es6.jss:471
e.evtKeydown @ index.es6.jss:355
e.handleEvent @ index.es6.jss:311
```

**Comment by banshee** [comment]
Sorry - I'm wrong, I was confusing two operations.  hitting return does restart the kernel.  It doesn't redraw the page though.  I'm not sure why I thought there was a difference between hitting return and clicking the button though; clearly that was just pilot error.

**Comment by devdanke** [comment]
> This works perfectly fine to resolve this issue.
> 
> from IPython.core.display import HTML HTML("<script>Jupyter.notebook.kernel.restart()</script>")
> 
> as stated over here https://stackoverflow.com/questions/37472274/jupyter-disable-restart-kernel-warning/ by kingbase

May work in Jupyter Notebook. Does not work in Jupyter Lab.  A comment on that suggested answer also confirms this command does not work on Jupyter Lab.  Also, this suggestion does not "resolve" the issue.  This suggestion is kludgy workaround.

**Comment by krassowski** [comment]
@devdanke if you are interested in making a contribution, I am happy to help.

1) For programmatic use, to implement my suggestion from https://github.com/jupyterlab/jupyterlab/issues/6420#issuecomment-920339993 one would need to check for presence of `args.ask` (`args.skipDialog`?) in:
    
    https://github.com/jupyterlab/jupyterlab/blob/5ea1bc727db6a79ecba78033a423d731d1d14837/packages/notebook-extension/src/index.ts#L1976-L1986
    
    and if `args.ask` is set to `false` (but not `undefined`!), call `current.context.sessionContext.restartKernel()` instead of `sessionDialogs!.restart`

    The same would need to be done for `console:restart-kernel'` in:
    
    https://github.com/jupyterlab/jupyterlab/blob/291ef37a98ce8f0455bb472485bb0a091b68b5c5/packages/console-extension/src/index.ts#L679-L692

2) Separately, we can implement a non-programmatic solution too: since [#12953](https://github.com/jupyterlab/jupyterlab/pull/12953) was merged last week, we now have a dedicated checkbox support in dialogs windows. We could have *Do not ask again in this session* or *Do not ask again for this notebook* checkbox added to the restart dialog which is defined in here:
   https://github.com/jupyterlab/jupyterlab/blob/5ea1bc727db6a79ecba78033a423d731d1d14837/packages/apputils/src/sessioncontext.tsx#L1353-L1360
    
    See https://jupyterlab.readthedocs.io/en/latest/extension/ui_helpers.html#generic-dialog

**Comment by krassowski** [comment]
Everyone here (especially @BoPeng who was first to suggest a checkbox), which of the checkbox implementations listed below do you think would be most useful?

- a) `[x] Do not ask again in this session`: restart dialogs silenced until user reloads JupyterLab
- b) `[x] Do not ask again for this notebook`: restart dialogs silenced for this specific notebook until user reloads JupyterLab
- c) `[x] Do not show this dialog ever again`: disables the dialog altogether, user needs to go to settings to re-enable it


I am personally against (c):
 - it could be just a setting rather than a checkbox in the dialog
 - if it is a checkbox, we need to explain "go to settings to reverse" or user will be angry if they click this accidentally

**Comment by michaelaye** [comment]
I'm for b) and/or a config setting

**Comment by rpm5099** [comment]
I think `b)` is reasonable, but `c)` should be available in the advanced config.  If someone loses the cure for cancer because they accidentally restarted their notebook and was expecting a confirmation dialogue, at least it saved the rest of us a some time and mouse clicks.

**Comment by sibosend** [comment]
hope this feature could be released ASAP, and option b looks good to me.

**Comment by krassowski** [comment]
@maxima120 this issue is closed because the `skipKernelRestartDialog` settings was released in JupyterLab 4.3

**Comment by sstraub-hmmc** [comment]
How does the `skipKernelRestartDialog` setting help with this issue? Is there a way now to restart the kernel by entering a command in a cell, without any user interaction required?

When I set `skipKernelRestartDialog=true` and then run `os._exit(0)`, I still get the "Kernel Restarting" modal.
